### PR TITLE
Edit dream

### LINKED ISSUE #3: Use case: edit dream
As a **User**,
I want to **edit** a **Dream**,
so that I can make some changes on its fields.

## Validation
- Title must be at most 40 chars long
- Description must be at most 200 chars long
- Title and description should not be null
- Date must not be in the future

### PR DESCRIPTION
**Pull request by Cotel**
### :pushpin: References
* **Issue:** close #3 
* **Related pull-requests:** #17 

### :tophat: What is the goal?

As a creator of a Dream I want to be able to edit that dream.

### :memo: How is it being implemented?

I've created a new use case for editing dreams which is a lot like `CreateDream`. I've done several refactors for reusing existing code and favouring a functional fashion 😎 🤓 

### :boom: How can it be tested?

Since updating is almost creating there are only two corner cases to test:

* User is not creator of dream
* Dream is not found

- [x] **Edit a Dream:**
  - [x] Create a Dream
  - [x] Update it (check #2 for validation corner cases)
  - [x] Try to update a Dream from another user
  - [x] Try to update a non existing Dream
 
### :floppy_disk: Requires DB migration?

- [x] Nope, we can just merge this branch.
- [ ] Yes, but we need to apply it before merging this branch.
- [ ] Yes, it's already applied.

### CODE REVIEW

File: src/main/kotlin/com/infusionvlc/somniumserver/dreams/usecases/EditDream.kt [hunk 1]
```diff
@@ -25,29 +27,52 @@ class EditDream(
...
-    .toEither { DreamEditionErrors.DreamNotFound(dreamId) }
-    .flatMap {
+  ): Either<DreamEditionErrors, Dream> = (
+    retrieveDreamFromPersistence(dao, dreamId) andThen
+      checkUserIsCreatorOfDream(userId) andThen
+      updateDream(dreamRequest, currentTime) andThen
+      validate(currentTime) andThen
+      storeDream(userId)
+    )()
+
```
Comment (Cotel) [inline src/main/kotlin/com/infusionvlc/somniumserver/dreams/usecases/EditDream.kt:37]: What about now? @tonilopezmr

**Review by Diphendara — APPROVED** [review]
Everythink ok! 👌

File: src/main/kotlin/com/infusionvlc/somniumserver/dreams/usecases/EditDream.kt (hunk 1 above)
Comment (Cotel) [reply src/main/kotlin/com/infusionvlc/somniumserver/dreams/usecases/EditDream.kt:37]: I'm interested on your opinions too @alvarocasadoc @Diphendara @kevinzs. Which way do you prefer? Do you understand this better now? We can work on making this even simpler but this is "functional" enough.

Comment (kevinzs) [reply src/main/kotlin/com/infusionvlc/somniumserver/dreams/usecases/EditDream.kt:37]: Now I think that can be a little easier for anyone who see functional code for the first time. Anyway we can progressively change the code to a more compact style.

Comment (Cotel) [reply src/main/kotlin/com/infusionvlc/somniumserver/dreams/usecases/EditDream.kt:37]: Bump

**Review by Diphendara — APPROVED** [review]

File: src/main/kotlin/com/infusionvlc/somniumserver/dreams/usecases/EditDream.kt (hunk 1 above)
Comment (Diphendara) [reply src/main/kotlin/com/infusionvlc/somniumserver/dreams/usecases/EditDream.kt:37]: Fuck, i delete my review :') but u already read it

Comment (Cotel) [reply src/main/kotlin/com/infusionvlc/somniumserver/dreams/usecases/EditDream.kt:37]: I didn't read any review for this code and I'm waiting for feedback 😢

Comment (Diphendara) [reply src/main/kotlin/com/infusionvlc/somniumserver/dreams/usecases/EditDream.kt:37]: I talk about the question, but i don't have any to say about the code, looks fine for me.

Comment (Cotel) [reply src/main/kotlin/com/infusionvlc/somniumserver/dreams/usecases/EditDream.kt:37]: Do you prefer it as it is now or do you like it like it was before? 

Furthermore, we can even change it into two more ways: 

Do you prefer to remove the `andThen` operators and work with extension functions instead of higher order functions (functions returning lambdas)?

```kotlin
fun execute(
    dreamId: Long,
    dreamRequest: DreamRequest,
    userId: Long,
    currentTime: Long = System.currentTimeMillis()
  ): Either<DreamEditionErrors, Dream> = retrieveDreamFromPersistence(dreamId)
      .checkUserIsCreatorOfDream(userId)
      .updateDream(dreamRequest, currentTime)
      .validate(currentTime)
      .storeDream(userId)
```
This style is the least functional (but maybe it is the more pragmatic)

Or do you prefer using more advanced concepts of functional programming and don't write code like a stream and do sequential operations?

```kotlin
fun execute(
    dreamId: Long,
    dreamRequest: DreamRequest,
    userId: Long,
    currentTime: Long = System.currentTimeMillis()
  ): Either<DreamEditionErrors, Dream> = ForEither extensions {
  bindings {
   val dream = retrieveDreamFromPersistence(dreamId).bind()
   
   if (isUserCreatorOfDream(userId, dream)) {
     val updatedDream = updateDream(dream, dreamRequest, currentTime).bind()
     validateDream(updatedDream, currentTime).bind()
   } else {
     UserIsNotCreatorOfDream.left()
   }
  }.fix()
}
```
This is the `monad comprehension` style and it's very _haskelly_ but it introduces some noise and it's not as flexible as pure function composition.

Comment (Diphendara) [reply src/main/kotlin/com/infusionvlc/somniumserver/dreams/usecases/EditDream.kt:37]: I like the 1º option of this comment, I can understand it and read more quickly

Comment (Cotel) [reply src/main/kotlin/com/infusionvlc/somniumserver/dreams/usecases/EditDream.kt:37]: More opinions on this please @alvarocasadoc @tonilopezmr @kevinzs @alexares @jodoldar @dredonieto

Comment (kevinzs) [reply src/main/kotlin/com/infusionvlc/somniumserver/dreams/usecases/EditDream.kt:37]: Choosing between this two options I prefer the first. I didn't like the third, I prefer the original form instead of this :laughing: .
